### ISSUE TITLE
different behavior when requesting a non-existent domain over HTTP and HTTPS. (affects 1.11.1 version)

### ISSUE THREAD

**Issue by ilias-sp** [tinyproxy/tinyproxy#441]
## Tinyproxy version
```
[root@pi4 tinyproxy] > tinyproxy -v
tinyproxy 1.11.1
[root@pi4 tinyproxy] > 
```

## Issue
I am adding custom html Errorfiles and i noticed a difference in behavior of tinyproxy when requesting an HTTP vs HTTPS site that does not exist:

- when requesting HTTP of an non existent domain, the Errorfile is triggered
- when requesting HTTPS of an non existent domain, the request passes through and browser displays its built-in "Server not found" page.

Shouldn't these two scenarios have the same handling from tinyproxy? From logs point of view (info level), i see the same lines besides that HTTP initiates a GET, while HTTPS its a CONNECT.

thank you

**Comment by ilias-sp** [comment]
its a duplicate of #440 but for latest version `1.11.1`

**Comment by ilias-sp** [comment]
hello!

so i installed squid and from firefox point of view, the behaviour is the same (HTTPS gets the page not found, HTTP shows some squid error page)

running from curl, the 2 requests are:

HTTP:

```
[ilias@yellowhat tmp] > curl -vvv -x localhost:3128 -v http://doesnotexistaaaaaaaaaa.gr    
*   Trying 127.0.0.1:3128...
* Connected to localhost (127.0.0.1) port 3128 (#0)
> GET http://doesnotexistaaaaaaaaaa.gr/ HTTP/1.1
> Host: doesnotexistaaaaaaaaaa.gr
> User-Agent: curl/7.79.1
> Accept: */*
> Proxy-Connection: Keep-Alive
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 500 Internal Server Error
< Server: squid/5.2
< Mime-Version: 1.0
< Date: Wed, 01 Jun 2022 12:23:20 GMT
< Content-Type: text/html;charset=utf-8
< Content-Length: 3854
< X-Squid-Error: ERR_CANNOT_FORWARD 0
< Vary: Accept-Language
< Content-Language: en
< X-Cache: MISS from yellowhat
< X-Cache-Lookup: MISS from yellowhat:3128
< Via: 1.1 yellowhat (squid/5.2)
< Connection: keep-alive
< 
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html><head>
<meta type="copyright" content="Copyright (C) 1996-2021 The Squid Software Foundation and contributors">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
```

HTTPS:

```
[ilias@yellowhat tmp] > curl -vvv -x localhost:3128 -v https://doesnotexistaaaaaaaaaa.gr -k
*   Trying 127.0.0.1:3128...
* Connected to localhost (127.0.0.1) port 3128 (#0)
* allocate connect buffer!
* Establish HTTP proxy tunnel to doesnotexistaaaaaaaaaa.gr:443
> CONNECT doesnotexistaaaaaaaaaa.gr:443 HTTP/1.1
> Host: doesnotexistaaaaaaaaaa.gr:443
> User-Agent: curl/7.79.1
> Proxy-Connection: Keep-Alive
> 
< HTTP/1.1 500 Internal Server Error
< Server: squid/5.2
< Mime-Version: 1.0
< Date: Wed, 01 Jun 2022 12:23:51 GMT
< Content-Type: text/html;charset=utf-8
< Content-Length: 3842
< X-Squid-Error: ERR_CANNOT_FORWARD 0
< Vary: Accept-Language
< Content-Language: en
< 
* Received HTTP code 500 from proxy after CONNECT
* CONNECT phase completed!
* Closing connection 0
curl: (56) Received HTTP code 500 from proxy after CONNECT
[ilias@yellowhat tmp] > 
```

googling it a bit i found this [SO thread](https://stackoverflow.com/questions/8608777/transparent-proxy-giving-ssl-error-when-accessing-via-https/8612660#8612660), and the guy says something i overlooked:


> SSL/TLS is endpoint security. This means that the data must be sent between the client and the server via the single encrypted channel. 

I guess that means that the proxies work as "pass through" in the HTTPS requests, not checking even if destination domain exists, and obviously not possible to return an error page if something bad occurred. (would be nice if at least they could check if domain exists or not, but oh well :) )

These are what i understand with my limited exposure to proxies. But if you agree, i guess we can close the issue.

**Comment by rofl0r** [comment]
> < Content-Length: 3842

what's interesting here is that curl discards the html body even though a content-length is sent. the spec https://httpwg.org/specs/rfc7231.html#CONNECT doesn't actually explicitly state what should happen in the error case, but it makes it clear that in the success case the proxy may not return a body. this means we're in the realm of "implementation defined behaviour" so the programs can do whatever they want.

**Comment by ilias-sp** [comment]
yes, looks like we cannot "redirect" or inject any html content when this scenario occurs, each client will do its own things. in `curl`s case, seem it cares about the HTTP 500 error and nothing else.

It would be nice if i could redirect to a custom error page as i do for regular HTTP requests, but seems not possible by design.

i will close the issue, we learnt a few things i guess from this :)
